runtime/runc_jailer.go

```diff
@@ -0,0 +1,370 @@
+			}
+
+			// Link the kernel image
+			newKernelImagePath := filepath.Join(rootPath, kernelImageFileName)
+			j.logger.WithField("newKernelImagePath", newKernelImagePath).Debug("linking kernel image")
+			if err := os.Link(m.Cfg.KernelImagePath, newKernelImagePath); err != nil {
+				return errors.Wrapf(err, "failed to link kernel image from path %v", m.Cfg.KernelImagePath)
+			}
+			m.Cfg.KernelImagePath = kernelImageFileName
+
+			// link drives to new contents path
+			for i, d := range m.Cfg.Drives {
+				drivePath := firecracker.StringValue(d.PathOnHost)
+				fileName := filepath.Base(drivePath)
+
+				if err := os.Link(drivePath, filepath.Join(rootPath, fileName)); err != nil {
+					return errors.Wrapf(err, "failed to copy drive %v", drivePath)
+				}
+
+				j.logger.WithField("drive", filepath.Join(rootPath, fileName)).Debug("Adding drive")
+				m.Cfg.Drives[i].PathOnHost = firecracker.String(fileName)
+			}
+
+			// Setting the proper path to where the vsock path should be
+			for i, v := range m.Cfg.VsockDevices {
+				filename := filepath.Base(v.Path)
+				v.Path = filepath.Join("/", filename)
+				m.Cfg.VsockDevices[i] = v
+			}
+
+			if err := filepath.Walk(rootPath, func(name string, info os.FileInfo, err error) error {
+				if err != nil {
+					return err
+				}
+
+				j.logger.WithField("path", name).Debug("chowning")
+
+				return os.Chown(name, int(j.uid), int(j.gid))
+			}); err != nil {
+				return errors.Wrap(err, "failed to chown directories")
+			}
+
+			j.logger.Info("Successfully ran jailer handler")
+			return nil
+		},
+	}
+}
+
+// BuildLinkFifoHandler will return a new firecracker.Handler with the function
+// that will allow linking of the fifos making them visible to Firecracker.
+func (j runcJailer) BuildLinkFifoHandler() firecracker.Handler {
+	return firecracker.Handler{
+		Name: jailerFifoHandlerName,
+		Fn: func(ctx context.Context, m *firecracker.Machine) error {
+			contentsPath := j.RootPath()
+			fifoFileName := filepath.Base(m.Cfg.LogFifo)
+			newFifoPath := filepath.Join(contentsPath, fifoFileName)
+			if err := os.Link(m.Cfg.LogFifo, newFifoPath); err != nil {
+				return err
+			}
+			m.Cfg.LogFifo = newFifoPath
+
+			metricFifoFileName := filepath.Base(m.Cfg.MetricsFifo)
+			newMetricFifoPath := filepath.Join(contentsPath, metricFifoFileName)
+			if err := os.Link(m.Cfg.MetricsFifo, newMetricFifoPath); err != nil {
+				return err
+			}
+			m.Cfg.MetricsFifo = newMetricFifoPath
+
+			return nil
+		},
+	}
+}
+
+// ExposeDeviceToJail will inspect the given file, srcFilePath, and based on the
+// file type, proper handling will occur to ensure that the file is visible in
+// the jail.
+func (j runcJailer) ExposeDeviceToJail(srcMountPath string) error {
+	uid := j.uid
+	gid := j.gid
+
+	stat := syscall.Stat_t{}
+	if err := syscall.Stat(srcMountPath, &stat); err != nil {
+		return err
+	}
+
+	path := j.RootPath()
+	dst := ""
+	if (stat.Mode & syscall.S_IFMT) == syscall.S_IFBLK {
```

Comment:
This needs comments explaining what and why.  "proper handling" in the godoc above isn't really sufficient.

Reply:
@xibz Can you address this comment?